Bamboo Service: Fix missing credentials & URL handling

bentolor opened the pull request:
_Update:_ Added another fix which fixes an error introduced in the latest changes using `URI.join()`  in favour to `URL.parse()` and rebased to the latest master.

This improves the Bamboo Service and provides two fixes:
1. One for the situation, where the build trigger won't work because Bamboo is requiring authentication credentials for the trigger GET: https://github.com/gitlabhq/gitlabhq/pull/9428/commits/8f25aca307b49ee006172b8c2985a878800aa6b6
2. One which fixes the way how the configured Bamboo base URL is assembled to the final REST URL. https://github.com/gitlabhq/gitlabhq/pull/9428/commits/fe9eb30d7ebe4a83eefea7e06f8b69b135dad15d
## Regarding credentials

The change now does provide additional HTTP Basic Auth parameters if user credentials were provided and appends an request parameter indicating the HTTP Basic Authentication should be used. This aligns interaction with Bamboo with the other calls this service executes.
## Regarding URL handling

If one had configured a `bamboo_url` like http://foo.bar/bamboo in the previous implementation the plugin directed it's request i.e. to http://foo.bar/rest/... instead of http://foo.bar/bamboo/rest/...

This i probably an unwanted side effect of how Ruby's `URI.join` is working. It will only work correctly, if 
- ... the prefix URL has at least one or more  trailing `/`
- .. the appendix parts are _not_ prefixed with `/`

Backported & tested successfully in our environment.

bentolor commented:
No clue why CI fails. Rebasing onto master didn't fix it. From my point of view the failure is probably caused by `master` or the CI environment.

stanhu commented:
@dblessing, can you review this one?

dblessing commented:
@bentolor Project services have a serialized configuration so you can create new/specific service configuration fields. I would rather see username/password implemented as separate fields than something parsed from a URI. That's the reason we moved to the serialized field, actually. Would you mind exploring that, please?

bentolor commented:
@dblessing I'm sorry, I don't understand. (Maybe due I'm a non-native speaker). 

The change simply duplicates your implementation of the [code block in `build_info(sha)`](https://github.com/gitlabhq/gitlabhq/commit/93bed4435e9e61570dd27e1660c230c3c323a744#diff-9aa8169ccafcf0ea06afa51130ab53eaR54) into `execute(_data)`. 

As far as I understand this patch does not involve any change to a existing configuration or field nor does it parse any URI. It simply attaches a single `&os_authType=basic` to the trigger URL and provides the configured credential values as basic auth parameters.

What 'serialized configuration' do you mean? Can you elaborate on this please?

dblessing commented:
@bentolor sorry for the confusion! please disregard my message.

dblessing commented on app/models/project_services/bamboo_service.rb:
```diff
@@ -131,7 +131,18 @@ def execute(data)
+    if username.blank? && password.blank?
+      self.class.get(url, verify: false)
+    else
+      get_url = "#{url}&os_authType=basic"
```
Did `os_authType=basic` work for you, still? What version of Bamboo? I was in the process of testing another issue with Bamboo service and had to remove this parameter for basic auth to work in 5.9.1.

bentolor commented:
Yes, `os_authType=basic` is required and does work as expected on Bamboo version 5.7.2 - 02 Dec 14. Without, the trigger URL only sees the login page.

But I do roughly remember, that in earlier versions Bamboo had another bug, and you'd need to pass the username and password *as additional query parameters*, though the pragma clearly states HTTP BASIC. So the URL had to look sth. like `?os_authType=basic&username=myname&password=mypass`. Maybe your Bamboo still requires this legacy format?

jblaine commented:
This would be hard for me to test here, but I ended up here only after trying for an hour to get Gitlab CE 7.14.1 to trigger Bamboo 5.9 builds. It doesn't work for me at all and I followed the service instructions verbatim.

So I guess this is just a +1

jvanbaarsen commented:
This merge request has been closed because a request for more information has not been reacted to for more than 2 weeks. If you respond and conform to the merge request guidelines in our [contributing guidelines](https://gitlab.com/gitlab-org/gitlab-ce/blob/master/CONTRIBUTING.md#pull-requests) we will reopen this merge request.

bentolor commented:
@jvanbaarsen I'm sorry - which request from my side is outstanding? I answered all requests. So what is open to do?

jvanbaarsen commented:
@bentolor I'll take a closer look at this PR later today. In the meanwhile I'll re-open

bentolor committed:
Honor credentials on calling Bamboo CI trigger

This improves the Bamboo Service and provides a fix for situations,
where the build trigger won't work, because Bamboo is requiring
authentication also for the trigger GET.

The change now does provide additional HTTP Basic Auth parameters
if user credentials were provided and appends an request parameter
indicating the HTTP Basic Authentication should be used.
This aligns interaction with Bamboo with the other calls this service
executes.

bentolor committed:
Fix broken URI joining for `bamboo_url` with suffixes

If one had configured a `bamboo_url` like http://foo.bar/bamboo in the
previous implementation the plugin directed it's request i.e. to
http://foo.bar/rest/... instead of http://foo.bar/bamboo/rest/...

`URI.join` only works correctly, if the prefix URL has
  - at least one or more  trailing '/'
  - the appended parts are _not_ prefixed with '/'

The current implementation should work with all sorts of Bamboo base URLs.

bentolor commented:
I rewrote/ported my patch based on the latest `master` and added another fix introduced by later changes (see updated issue description).

Would be very happy if this change would make its way into upstream so we can stop hot-patching our installations. 

Sadly the issue, that the merge status is still not working correctly for branches will still not work. But at least the trigger works. There happened to be very much discussion about the branch status issue in https://gitlab.com/gitlab-org/gitlab-ce/issues/1355 and  https://gitlab.com/gitlab-org/gitlab-ce/merge_requests/2562 though that one is lost as the author retracted his branch. And Atlassians REST API lacks important functions to tackle that correctly.

connorshea commented:
@bentolor I'm really sorry for the inconvenience here, but we're no longer accepting PRs from the GitHub repository because we've not been able to keep up with two separate projects at once.

I'd love to have you open this as a merge request over at [the GitLab.com project](https://gitlab.com/gitlab-org/gitlab-ce/merge_requests). You can log in on GitLab.com using your GitHub account, and then fork the CE repository, make it a remote for your local copy of the GitHub repo, and then push to it and open an MR.

If you'd prefer, I can open the MR myself and credit you, although if there are any problems I likely won't be able to fix them myself.

Sorry for the extra work this'll cause, and sincerely, thank you for taking the time to contribute back upstream.

bentolor commented:
Thank you @connorshea for your kind and detailed feedback!

Closing this request in favour to my new [merge request at Gitlab](https://gitlab.com/gitlab-org/gitlab-ce/merge_requests/4367). I already assumed that meanwhile Gitlab.com is the better place for getting things into Gitlab.com
